File: backend-java/src/main/java/com/empire/employeefinder/service/EmployeeService.java
package com.empire.employeefinder.service;

import com.empire.employeefinder.dto.request.EmployeeFilterRequestDto;
import com.empire.employeefinder.dto.request.EmployeeRequestDto;
import com.empire.employeefinder.dto.response.EmployeeResponseDto;
import com.empire.employeefinder.model.enums.Gender;
import com.empire.employeefinder.model.enums.Location;
import com.empire.employeefinder.model.enums.Status;

import java.io.IOException;
import java.security.Principal;
import java.util.List;

public interface EmployeeService {

    EmployeeResponseDto add(EmployeeRequestDto employeeRequestDto, Principal principal) throws IOException;

    EmployeeResponseDto update(Long employeeId, EmployeeRequestDto employeeRequestDto, Principal principal) throws IOException;

    void changeEmployeeStatus(Long employeeId, Status newStatus, String login);

    EmployeeResponseDto getById(Long employeeId);

    void delete(Long employeeId, Principal principal);

    List<EmployeeResponseDto> filterEmployees(EmployeeFilterRequestDto dto, String sortField, String sortDirection,
                                              int pageSize, int pageNumber, String searchField, String parameter);

    List<EmployeeResponseDto> filterByJobTypeIdAndPositionIdWithSearch(
            Long jobTypeId, Long jobPositionId, Gender gender, Location location,
            String searchField, String parameter,
            String sortField, String sortDirection,
            int pageSize, int pageNumber);

    List<EmployeeResponseDto> getAll(String searchField, String parameter, String status,
                                     String sortField, String sortDirection,
                                     int pageSize, int pageNumber);

    List<EmployeeResponseDto> getAllForUser(String searchField, String parameter,
                                            String sortField, String sortDirection,
                                            int pageSize, int pageNumber);
}

File: frontend-react/src/services/EmployeeService.js
import axios from 'axios';

const EMPLOYEES_URL = "http://localhost:8081/api/v1/employees";

class EmployeeService {

  addEmployee = (formData) => {
    return axios.post(`${EMPLOYEES_URL}/add`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
  };

  updateEmployee = (employeeId, formData) => {
    return axios.put(`${EMPLOYEES_URL}/update/${employeeId}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
  };

  deleteEmployee = (employeeId) => {
    return axios.delete(`${EMPLOYEES_URL}/remove/${employeeId}`);
  };

  changeEmployeeStatus = (employeeId, newStatus) => {
    return axios.put(`${EMPLOYEES_URL}/change-status/${employeeId}`, null, {
      params: {
        status: newStatus,
      }
    });
  };

  getEmployeeById = (id) => {
    return axios.get(`${EMPLOYEES_URL}/${id}`);
  };

  getAllSortedEmployeesByPages = (sortField, pageSize, pageNumber, sortDirection = "asc") => {
    return axios.get(`${EMPLOYEES_URL}?sortField=${sortField}&sortDirection=${sortDirection}&pageSize=${pageSize}&pageNumber=${pageNumber}`);
  };

  getAllSortedEmployeesSearched = (searchField, parameter, sortField, pageSize, pageNumber, sortDirection = "asc") => {
    return axios.get(`${EMPLOYEES_URL}?searchField=${searchField}&parameter=${parameter}&sortField=${sortField}&sortDirection=${sortDirection}&pageSize=${pageSize}&pageNumber=${pageNumber}`);
  };

  getAllSortedEmployeesByPagesForUser = (sortField, pageSize, pageNumber, sortDirection = "asc") => {
    return axios.get(`${EMPLOYEES_URL}/for-user?sortField=${sortField}&sortDirection=${sortDirection}&pageSize=${pageSize}&pageNumber=${pageNumber}`);
  };

  getAllSortedEmployeesSearchedForUser = (searchField, parameter, sortField, pageSize, pageNumber, sortDirection = "asc") => {
    return axios.get(`${EMPLOYEES_URL}/for-user?searchField=${searchField}&parameter=${parameter}&sortField=${sortField}&sortDirection=${sortDirection}&pageSize=${pageSize}&pageNumber=${pageNumber}`);
  };

  filterEmployees = (filterDto, sortField, sortDirection = "asc", pageSize = 25, pageNumber = 0) => {
    const params = {
      sortField,
      sortDirection,
      pageSize,
      pageNumber,
    };

    return axios.post(`${EMPLOYEES_URL}/filter`, filterDto, { params });
  };

  filterEmployeesSearched = (
    filterDto,
    sortField,
    sortDirection = "asc",
    pageSize = 25,
    pageNumber = 0,
    searchField,
    parameter
  ) => {
    const params = {
      sortField,
      sortDirection,
      pageSize,
      pageNumber,
      searchField,
      parameter
    };

    return axios.post(`${EMPLOYEES_URL}/filter`, filterDto, { params });
  };

  filterByJobTypeAndPosition = (jobTypeId, jobPositionId, page = 0, size = 20, sortBy = "id", sortDir = "asc") => {
    return axios.get(`${EMPLOYEES_URL}/filter/job-type-position`, {
      params: { jobTypeId, jobPositionId, page, size, sortBy, sortDir }
    });
  }
}

export default new EmployeeService();